[WIP] StatusBar, Dashboard and Provider UI improvements

arturi opened the pull request:
This PR:

- Moves `Upload` button into the StatusBar itself, going from “cloud icon only” to plain simple text;
- Adds `Upload +3 files` button for when you pick more files while upload is in progress;
- Adds roles `presentation`, `alert` and `progressbar` for a11y, turn button without action into a block, and more;
- Adds `Done` checkmark, similar to when you are editing a file metadata, to Provider view to signal that you finished picking files from Instagram;
- Trying to remove white backgrounds from FileItems; less red;

Many of these new UI suggestions are coming from @nqst 🥇

ToDo:

- [x] nicer checkmark icons and better placement in Provider views
- [x] ability to hide upload button in StatusBar (we have this option in Dashboard at the moment)

<img width="852" alt="screen shot 2017-11-25 at 8 13 32 pm" src="https://user-images.githubusercontent.com/1199054/33273567-ad90d8bc-d35b-11e7-85b9-cda1ca79a6d1.png">

<img width="920" alt="screen shot 2017-11-26 at 10 58 52 pm" src="https://user-images.githubusercontent.com/1199054/33273569-af14f9f2-d35b-11e7-84e2-e7dd6986a47e.png">

<img width="891" alt="screen shot 2017-11-27 at 10 18 13 am" src="https://user-images.githubusercontent.com/1199054/33273805-5e33472c-d35c-11e7-8093-7df4bfe22d57.png">

<img width="852" alt="screen shot 2017-11-27 at 10 18 29 am" src="https://user-images.githubusercontent.com/1199054/33273808-6169c2f4-d35c-11e7-9be1-0aa588d3dd1a.png">

Closes #255 by renaming “Authenticate” to “Connect”.

goto-bus-stop commented:
🎉 🎉  exciting! this looks great!

nqst commented:
Great job! 👍

arturi commented:
Checkboxes for grid/list views. @ifedapoolarewaju you had some comments on this too, so welcome here :)

<img width="847" alt="screen shot 2017-11-27 at 2 35 27 pm" src="https://user-images.githubusercontent.com/1199054/33285584-bd198c0a-d380-11e7-87c0-28fea33395e2.png">

<img width="842" alt="screen shot 2017-11-27 at 2 35 19 pm" src="https://user-images.githubusercontent.com/1199054/33285586-bfd79bc6-d380-11e7-8c85-778d95b329ca.png">

ifedapoolarewaju commented:
@arturi well now that I see this lovely design you've got going on, I have no comments 😃

goto-bus-stop commented:
is it supposed to have both buttons? 😄 

![image](https://user-images.githubusercontent.com/1006268/33526256-b829b2c0-d83e-11e7-9296-6a8f0dc4e828.png)

i think it could make sense to move this Upload button to the right, since it is also on the right of the statusbar when it is visible.

goto-bus-stop commented:
> We could make it so that on desktop these empty circles only appear on :hover.

I don't mind always showing them personally.

goto-bus-stop commented on src/plugins/StatusBar/StatusBar.js:
```diff
@@ -106,24 +106,58 @@ module.exports = (props) => {
+                      onclick=${props.retryAll}>
+                ${props.i18n('retry')}
+              </button>`
+}
```
I wonder if we could pass these actions in as an option somehow? for a standalone StatusBar, not inside the Dashboard, having the upload button is a bit strange.

maybe so that the dashboard can do something like

```js
this.core.use(StatusBar, {
  id: `${this.id}-StatusBar`,
  actions (state) {
    return [ hasNewFiles && UploadBtn({ /* xyz */ }), hasError && RetryBtn({ /* xyz */ }) ]
  }
})
```

goto-bus-stop replied:
or more simply, we could also have a boolean option that toggles the visibility of the action buttons.

arturi replied:
I’ve added an option to hide the Upload button, do you think similar is needed for each of the other buttons or all of them?

> for a standalone StatusBar, not inside the Dashboard, having the upload button is a bit strange

Is is though? You have a DragDrop area, stick StatusBar underneath, and now you have a functioning uploader! :-)

goto-bus-stop commented on src/plugins/StatusBar/StatusBar.js:
```diff
@@ -106,24 +106,58 @@ module.exports = (props) => {
+      <div role="progressbar"
+           aria-valuemin="0"
+           aria-valuemax="100"
+           aria-valuenow="${value}">
```
when `props.mode` is 'indeterminate', we should not add the `aria-valuenow` attribute

arturi replied:
And all of them, including min/max?

arturi replied:
I also think this would result in duplicated `aria-value`s, since the root element has them? Removed on `ProgressBarProcessing`.

goto-bus-stop replied:
you can sort of do conditional attributes by using object spread, both in jsx and hyperx:

```js
html`
  <div
    ${props.mode === 'determinate'
      ? { 'aria-valuenow': value }
      : {}}
  />
`
```

but i think the setAttribute is an acceptable solution too.

arturi replied:
Oh, nice, I didn’t know that! Thanks!

Do we need to check for `props.mode` if we can check for `progressValue`? If `in determinate`, won’t `progressValue` be 0 or undefined? I got lost here, sorry.

goto-bus-stop commented on src/plugins/StatusBar/StatusBar.js:
```diff
@@ -136,16 +170,16 @@ const ProgressBarUploading = (props) => {
 const ProgressBarComplete = ({ totalProgress, i18n }) => {
   return html`
-    <div class="UppyStatusBar-content">
+    <div class="UppyStatusBar-content" role="alert">
```
`alert` is a role for very important information, i think the progress is not _that_ important—would `status` be more appropriate?

arturi replied:
But would you say alert is fine for errors though?

goto-bus-stop commented on src/generic-provider-views/Browser.js:
```diff
@@ -60,6 +60,15 @@ module.exports = (props) => {
+      <button class="UppyButton--circular UppyButton--blue Browser-doneBtn"
+              type="button"
+              aria-label="Done picking files"
+              title="Done picking files"
```
can we localise these? not sure if we have easy access to a translator instance from `generic-provider-views`

arturi replied:
Yeah, that’s the problem, where would the strings come from? Dashboard makes the most sense, and at this point we basically have to admit that “target: Dashboard” will be the only way to use GoogleDrive and Instagram.

goto-bus-stop commented on src/plugins/StatusBar/StatusBar.js:
```diff
@@ -106,24 +106,59 @@ module.exports = (props) => {
+           role="progressbar"
+           aria-valuemin="0"
+           aria-valuemax="100"
+           aria-valuenow="${progressValue}"></div>
```
here `progressValue` can be null or undefined if we are processing and the processing progress is `'indeterminate'`, we should not add the valuenow attribute in that case

arturi commented:
Updated some icons (remove was buggy around the edges, text icon was terrible, this one is a little better), paddings, removed excess Upload button:

<img width="884" alt="screen shot 2017-12-03 at 3 35 33 pm" src="https://user-images.githubusercontent.com/1199054/33529503-c887bb34-d83f-11e7-9d55-e60271a23fd7.png">

arturi commented:
> i think it could make sense to move the new Upload button to the right, since it is also on the right of the statusbar when it is visible.

You mean when there’s progress its on the right side, so it makes sense to keep it their always?

Tried, I think its better on the right, but could be just that I’m used to seeing it that way by now, we can try it on the right too:

<img width="900" alt="screen shot 2017-12-03 at 3 36 11 pm" src="https://user-images.githubusercontent.com/1199054/33529533-1f88f5f6-d840-11e7-91a2-511d0a319718.png">

goto-bus-stop commented:
> You mean when there’s progress its on the right side, so it makes sense to keep it their always?

yeah! that last screenshot looks great to me 😸

arturi commented:
<img width="866" alt="screen shot 2017-12-03 at 3 58 58 pm" src="https://user-images.githubusercontent.com/1199054/33529730-f43d2a86-d842-11e7-98b1-e4322d8a317a.png">

arturi commented:
Didn’t want to prolong this, open to more comments and changes in future PRs or master 👌
